### DISCUSSION TITLE
Running ord

### DISCUSSION THREAD

**Discussion by openoms** [raspiblitz/raspiblitz#3900] [Show and tell]
Controversial as it is there are repeated questions about how to run Ord - https://github.com/casey/ord on the Raspiblitz.

I am testing on a amd64 server with 32 GB RAM  as according to community reports a raspberry pi would be underpowered for the initial sync. 

Also rpcuser and password auth with bitcoind is not yet supported so expect to break some other apps (LND, Specter, electrum servers, Mempool etc) configured to use that method: 
rpc username an password auth support WIP in ord: https://github.com/casey/ord/issues/1399
notes about setting apps to use rpc username an password: https://github.com/rootzoll/raspiblitz/issues/3691

Steps taken to install on a server running Raspiblitz on a Debian11 base (Ubuntu should be similar):
```
## INSTALL
# https://github.com/casey/ord#installation

# system dependencies
sudo apt-get install libssl-dev gcc -y

# create user
sudo adduser --disabled-password --gecos "" ord
# switch to user
sudo su - ord
# install rust for the user
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
source "$HOME/.cargo/env"
# download source
git clone https://github.com/casey/ord.git
cd ord
# build
cargo build --release

# prepare the data directory
# exit user
exit
sudo mkdir /mnt/hdd/app-storage/ord
sudo chown -R ord:ord /mnt/hdd/app-storage/ord

# prepare bitcoind (apps realying on rpcuser and rpcpassword auth will stop working)
# comment out rpcuser and rpcpassword in bitcoin.conf
sudo sed -i "s/^rpcuser/#rpcuser/g" /mnt/hdd/bitcoin/bitcoin.conf
sudo sed -i "s/^rpcpassword/#rpcpassword/g" /mnt/hdd/bitcoin/bitcoin.conf
# restart bitcoind
sudo systemctl restart bitcoind
# make the .cookie file group readable
sudo chmod 640 /mnt/hdd/bitcoin/.cookie
# add the ord user to the bitcoin group
sudo usermod -a -G bitcoin ord
```
```
# Build the ord database
# switch to user
sudo su - ord

# run ord (in tmux) to create the database with `--index-sats`
./ord/target/release/ord \
  --cookie-file /mnt/hdd/bitcoin/.cookie \
  --data-dir /mnt/hdd/app-storage/ord \
  --bitcoin-data-dir /mnt/hdd/bitcoin \
  --index-sats \
  index
```

some other things are running here, but the base load was below 1.5. Ord makes the machine work quite a bit. It progresses quickly until the inscriptions start:
![image](https://user-images.githubusercontent.com/43343391/221175882-adb242be-9809-4010-9bb6-04f9c757ff1b.png)
All the idexing is finished in an hour.

alias to be set for admin so you can just type `ord`:
```
alias ord="sudo -u ord /home/ord/ord/target/release/ord --cookie-file /mnt/hdd/bitcoin/.cookie --data-dir /mnt/hdd/app-storage/ord --bitcoin-data-dir /mnt/hdd/bitcoin"
```
Can paste it in:
`/home/admin/_aliases`
to persist.

Run the server in tmux (using the alias):
```
# allow the port through the firewall
sudo ufw allow 8000 comment ord-http
# run the server
ord server --http-port 8000
```

**Comment by mitohund** [comment]
(EDIT: added a fourth question)

I don't have a server to test this on, unfortunately. And I'm hesitant to use my RB since I need it to work reliably. I have been reading up on this a fair bit though. Here are a few things I noticed. Mostly from the perspective of me wanting to tinker around with this on my RB, not on a server. Which means I ran into three main questions:

1. How to install ord on a RB?
2. Is it possible to index on a RB?
3. If not, can an index file be copied to a RB?
4. Once the above has been sorted, does ord run well on RB or is all this effort in vain?

Here's what I found out so far:

1. Yes, by commenting out rpcuser and rpcpassword in bitcoin.conf, as described by @openoms above and somebody else in the Telegram channel before. This may break some other services one may or may not be using. There's [an issue](https://github.com/casey/ord/issues/1399) regarding fixing this, which seems to be "[a priority!](https://github.com/casey/ord/issues/527#issuecomment-1413417693)" Alternatively, somebody [modified ord](https://github.com/casey/ord/issues/1808#issuecomment-1439092708) to get it to run on RB. I don't fully understand what they did, so maybe somebody here could comment?
2. @openoms is testing this at the moment?
3. There seem to be lots of people copying the index between nodes (good discussion [here](https://discord.com/channels/987504378242007100/1076920434861281450)). There's even a torrent with somebody's index that other people use. A few things I noticed in this context. There seems to have been a breaking change in database format introduced in 0.5.0. And then in 0.5.1 indexing seems to have been made a lot more efficient both regarding time and filesize (down from 31GB to 2GB, even 767MB after compression). Some people (including @openoms) are finishing the entire task of indexing within an hour now. I'm hoping this means indexing has now become possible on a RB and am looking forward to @openoms results!
4. Looking forward to finding this out!

**Comment by openoms** [comment]
> How to install ord on a RB?

Manually for now following the commands described above. The same steps work on the RPi as on the server

> Is it possible to index on a RB?

Seems like it, but will take multiple hours of high load,

>If not, can an index file be copied to a RB?

Should be able to, but did not test.
Someone could try with the latest version and report.

> Once the above has been sorted, does ord run well on RB or is all this effort in vain?

Once the indexing is finished Ord is not heavier than electrs for example, it should run well. Can also be just started on demand.

Would suggest working in tmux for now to keep the processes running. See a simple guide: https://linuxize.com/post/getting-started-with-tmux/ or follow any other.

Overall the main problem is the cookie auth not being compatible with mempool and need to reconfigure the other apps. If you are not compfortable with the linux command line and  relying on anything apart from bitcoind and maybe CLN on the Raspiblitz it is not yet a good idea to hack around with ord.

**Comment by mitohund** [comment]
Thanks for all the info! I'm fine with the Linux command line. There are a few things I haven't done yet (like build something myself or tmux), but that's the main reason I like to play around with the RB and ord is giving me an excuse to do that. Will read up on tmux, thanks for the link.

I'm only using a few services on my RB (mempool.space, etc.) and could do without them temporarily. It should be possible to re-enable rpcuser and rpcpassword in bitcoin.conf, right? ord should then stop working, but mempool.space should work again?

> Is it possible to index on a RB?
Seems like it, but will take multiple hours of high load.

Could you leave a note here once your RB is through? I read the indexing can slow down massively towards the end. I could then plan ahead and schedule in the other services not working.

**Comment by openoms** [comment]
One for sure using  `ord --index-sats  index` - so you can check the ordinal values of sats as well - takes much longer. Probably a day on a server and multiple days on the RPi.

**Comment by heronimoo** [comment]
Does the RB Bisq service require rpcuser/rpcpassword or would it still work with those commented out in bitcoin.conf?

> One for sure

Did you mean "one day for sure"?

**Comment by openoms** [comment]
@heronimoo with the --index-sats setting my server is like half-way now, there is no hope to build this on an RPi. Will let you know the rough estimate once finished.
Might test copying over to the RPi, but I presume it can have problems with just keeping also.

**Comment by heronimoo** [comment]
@openoms I think you meant to respond to another poster above. But I guess I can ask follow-ups anyway ;) -> could an RPi keep up with the non-sats index? Would you be prepared to share your index file with somebody who doesn't have access to a massive server?

My previous question was actually this one - I would still be very interested in the answer, if you could:
> Does the RB Bisq service require rpcuser/rpcpassword or would it still work with those commented out in bitcoin.conf?

**Comment by openoms** [comment]
>Could an RPi keep up with the non-sats index? 

Yes, without the sats index probably.

>Would you be prepared to share your index file with somebody who doesn't have access to a massive server?

Not planning to share the index as it would need to be kept up to date. A used server like mine is like £120 BTW.  A laptop or desktop also works.

>Does the RB Bisq service require rpcuser/rpcpassword or would it still work with those commented out in bitcoin.conf?

yes, not relevant. If bitcoind is running can be connected to by Bisq. That is not an authenticated connection.

**Comment by heronimoo** [comment]
@openoms Thanks so much!

**Comment by mitohund** [comment]
@openoms did your RB ever manage to create the regular index on its own (or is it still running a few days later, poor little thing)?

**Comment by openoms** [comment]
> @openoms did your RB ever manage to create the regular index on its own (or is it still running a few days later, poor little thing)?

Nope, stopped it third of the way when realized it only gets the inscriptions, no sats index.
Might restart next week when moved my signet lightning node somewhere else. Priorities.

**Comment by mitohund** [comment]
Of course. Thanks for the update!

EDIT: Did you get the feeling that building the inscriptions index is feasible as opposed to the sats-index (which appears to be out)? Would we be talking hours, days, weeks?

**Comment by mitohund** [comment]
@openoms I followed your above instructions and they worked beautifully. My RB is indexing now. Let's see how far I get!

The only hiccup was that when I typed "ord server --http-port 8000", I received "ord is not in the sudoers file. This incident will be reported."

I launched ord with the previous command, so that works. But how do I fix the alias?

**Comment by openoms** [comment]
@mitohund the alias above is made for the admin user which can use sudo to run the command as the ord user., so you don't need to switch users that way.

**Comment by mitohund** [comment]
@openoms oh man I actually realised that in bed last night. Sometimes one just needs to get off the computer... Thanks anyway!

The index was built over night on my Pi. So that's actually totally fine. Will start testing today.

Last question, I hope: I would have expected to reach my ordinals server using a webbrowser and typing in my RB's IP plus port 8000. But I can't. Am I misunderstanding something?

**Comment by openoms** [comment]
yes, you should be able to access it when the server is running.

make sure that the firewall allowas that port through:
```
sudo ufw allow 8000 comment ord-http

**Comment by mitohund** [comment]
Ha! That did it. Thanks!

Maybe you should add this to your instructions above, for the noobs like me. An additional line for users who try this on an RB instead of a different server that they configured themselves.

**Comment by mitohund** [comment]
After each restart, I need to redo `sudo chmod 640 /mnt/hdd/bitcoin/.cookie`. Is there a way to automate that?

Also, how do I get the ord server to autostart?

**Comment by openoms** [comment]
> After each restart, I need to redo `sudo chmod 640 /mnt/hdd/bitcoin/.cookie`. Is there a way to automate that?

would need to change bitcoind.service, it is described in https://github.com/rootzoll/raspiblitz/issues/3691#issue-1598817368


> Also, how do I get the ord server to autostart?

Would need to make a systemd service for it too.
See the example how it is done for electrs in the electrs install script or if you have it installed:
`sudo systemctl cat electrs`

**Comment by mitohund** [comment]
Thank you!

**Comment by woiwoiwoiwoi** [comment]
Hi, another ambitious amateur here, who followed this info a few months ago and set up ord. Then life kept me busy for a few months. And now that I've returned to this, I noticed that a few things have changed and the above instructions are not correct anymore. I would appreciate some pointers as to how to proceed. Here's what I noticed:

1. The repo seems to have moved to https://github.com/ordinals/ord > Question: how can I update to a current version of ord now?
2. rpc username and password auth support [seems implemented](https://github.com/ordinals/ord/pull/1527) > does that mean my local mempool.space and ord can run at the same time now? How do I set this up?

**Comment by woiwoiwoiwoi** [comment]
I'll try to respond to my own questions and would appreciate comments if what I did was not the proper way. I really am just feeling my way forward.

```
## adapt to repo change & update ord
sudo su - ord
# if required, create backup of old ord folder
mv ord ord_bak
# download source
git clone https://github.com/ordinals/ord.git
# build
cd ord
cargo build --release

## exit user
exit

## I decided to rebuild the index from scratch
cd /mnt/hdd/app-storage/ord
# if required, create backup of old index  - don't remember if sudo was needed
mv index.redb index.redb.bak
# if no backup required, just delete the old index - don't remember if sudo was needed
rm index.redb

## switch back to rpcuser and rpcpassword auth, so that mempool etc. work again
sudo nano /mnt/hdd/bitcoin/bitcoin.conf
# uncomment rpcuser and rpcpassword
# note down rpcpassword if you don't have your Raspiblitz Password B handy
# Ctrl-X > y > Enter
# adapt alias (remove --cookie-file, add --bitcoin-rpc-user & --bitcoin-rpc-pass)
nano /home/admin/_aliases
# replace current ord alias with the following, make sure to add in your Raspiblitz Password B
alias ord="sudo -u ord /home/ord/ord/target/release/ord --bitcoin-rpc-user raspibolt --bitcoin-rpc-pass <Raspiblitz Password B> --data-dir /mnt/hdd/app-storage/ord --bitcoin-data-dir /mnt/hdd/bitcoin"
# I added a second alias to launch the server, make sure to add in your Raspiblitz Password B
alias ord-server="sudo -u ord /home/ord/ord/target/release/ord --bitcoin-rpc-user raspibolt --bitcoin-rpc-pass <Raspiblitz Password B> --data-dir /mnt/hdd/app-storage/ord --bitcoin-data-dir /mnt/hdd/bitcoin server --http-port 8000"

## restart
# either restart just bitcoind
sudo systemctl restart bitcoind
# or the entire Raspiblitz (so the aliases persist)
menu
reboot

## run the server to create index and provide web UI
ord-server
```

Indexing has been running for two days now and is still far from finished. Load is very high, but ord appears more stable (v 0.8.1) than before (v 0.5.2). Will report back once it's done.
